### PR TITLE
BF16 accuracy/performance mode support in HALO/ODLA

### PR DESCRIPTION
**Pull request by dj176050**
Add bf16 mode support in HALO; 
Remove unused code; 
Add gemm accuracy mode bf16 support; 
Add odla_argmax;
Fix some issues;

### CODE REVIEW

File: ODLA/include/ODLA/odla_common.h [hunk 1]
```diff
@@ -134,6 +134,14 @@ typedef enum {
...
 } odla_status;
 
+//! \brief BF16 mode
+typedef enum {
+  BF16_DISABLE,
+  BF16_ACCURACY_MODE,
+  BF16_PERFORMACE_MODE,
+  BF16_AUTO_MODE,
+} odla_bf16_mode;
+
```
Comment (weimingzha0) [inline ODLA/include/ODLA/odla_common.h:144]: This is execution related and is used together with "ODLA_BF16_MODE"
 how about define in "odla_compute.h" ?

File: lib/target/generic_cpp/generic_cxx_codegen.cc [hunk 2]
```diff
@@ -566,6 +566,21 @@ void GenericCXXCodeGen::RunOnFunction(Function& function) {
         os_ << "odla_SetComputationItem(Comp, ODLA_OPT_BATCH_SIZE, "
                "(odla_item_value) &opt_batch_size);\n";
       }
+      if (opts_.bf16_mode == CodeGen::BF16Mode::Accuracy) {
```
Comment (weimingzha0) [inline lib/target/generic_cpp/generic_cxx_codegen.cc:569]: can we make a common helper function like "emitBF16Mode" for line line 569-583 and line 625-639?

File: lib/target/generic_cpp/generic_cxx_codegen.cc [hunk 3]
```diff
@@ -566,6 +566,21 @@ void GenericCXXCodeGen::RunOnFunction(Function& function) {
...
+        os_ << "odla_bf16_mode mode = BF16_ACCURACY_MODE;\n";
+        os_ << "odla_SetComputationItem(Comp, ODLA_BF16_MODE, "
```
Comment (weimingzha0) [inline lib/target/generic_cpp/generic_cxx_codegen.cc:571]: line 571, 575, 579 are the same.
How about something like 

os_ << "odla_bf16_mode mode = " << get_bf16_mode(opts_bf16_mode) << ";\n";
os_ << "odla_SetComputationItem(Comp, ODLA_BF16_MODE, (odla_item_value) &mode);\n"

File: driver/driver.cc [hunk 4]
```diff
@@ -117,9 +117,16 @@ static llvm::cl::opt<bool> RISCVOpt(
...
 
-static llvm::cl::opt<bool> EnableBF16("enable-bf16",
-                                      llvm::cl::desc("Enable BF16"),
-                                      llvm::cl::init(false));
+static llvm::cl::opt<CodeGen::BF16Mode> BF16Mode(
+    llvm::cl::values(
+        clEnumValN(CodeGen::BF16Mode::Disable, "disable", "disable bf16 mode"),
+        clEnumValN(CodeGen::BF16Mode::Accuracy, "accuracy", "white list Model"),
+        clEnumValN(CodeGen::BF16Mode::Performace, "performace",
+                   "global enable bf16,excepte black list"),
```
Comment (weimingzha0) [inline driver/driver.cc:125]: "excepte" -> "except"

File: ODLA/platforms/dnnl/odla_dnnl.cc [hunk 5]
```diff
@@ -1360,116 +1406,86 @@ odla_value odla_ReduceMean(odla_value input, odla_size_t num_of_axes,
...
-  InterpretIfNeeded();
-
-  return is_elements_add ? odla_Add(v, bias, id) : v;
-}
-
-odla_value batch_gemm_op(odla_value lhs, odla_bool transpose_lhs,
-                         odla_value rhs, odla_bool transpose_rhs,
-                         odla_float32 alpha, odla_float32 beta, odla_value bias,
-                         odla_value_shape output_dims, const odla_value_id id) {
+odla_value odla_Gemm(odla_value lhs, odla_bool transpose_lhs, odla_value rhs,
```
Comment (weimingzha0) [inline ODLA/platforms/dnnl/odla_dnnl.cc:1409]: after "gemm_op" is removed, can we still support "batch_gemm_op"?

File: ODLA/include/ODLA/odla_common.h (hunk 1 above)
Comment (dj176050) [reply ODLA/include/ODLA/odla_common.h:144]: ok,  better to put in compute.h

File: ODLA/platforms/dnnl/odla_dnnl.cc (hunk 5 above)
Comment (dj176050) [reply ODLA/platforms/dnnl/odla_dnnl.cc:1409]: sure, i fused the batch_gemm and gemm function together

File: lib/target/generic_cpp/generic_cxx_codegen.cc [hunk 6]
```diff
@@ -101,6 +101,28 @@ static void EmitBanner(std::ostream* os, std::ostream* header_os,
   *os << "#include " << GetIncludeFile(api) << "\n\n";
 }
 
+static std::string GetBF16Mode(CodeGen::BF16Mode mode) {
+  std::ostringstream ss;
+  switch (mode) {
+    case CodeGen::BF16Mode::Accuracy: {
+      ss << "BF16_ACCURACY_MODE";
```
Comment (weimingzha0) [inline lib/target/generic_cpp/generic_cxx_codegen.cc:108]: just return the string?

**Review by weimingzha0 — APPROVED** [review]

**Pull request merged**